### ISSUE TITLE
Libtorch C++ load model error on windows10

### ISSUE THREAD

**Issue by hset911** [pytorch/pytorch#15919]
OS: WIN10
CUDA 10
CUDNN 7.3
PYTORCH 1.0

I trained unet. It works fine in a python environment.
I want to make it work in c++ by libtorch.
first,I use Official example to test.
Like this:
`import torch`
`import torchvision`

`model = torchvision.models.resnet50(pretrained=True)`
`model.eval()`

`example = torch.zeros(1, 3, 224, 224)`

`traced_script_module = torch.jit.trace(model, example)`
`traced_script_module.save("model.pt")`

then run it in cpp,like this:

`torch::DeviceType device_type = at::kCUDA;`
`std::cout << "use cuda" << std::endl;`
 `std::shared_ptr<torch::jit::script::Module> module = torch::jit::load(model_file);`
 `module->to(device_type);`
 `assert(module != nullptr);`
 `std::cout << "ok\n";`
 `std::vector<torch::jit::IValue> inputs;`
 `inputs.push_back(torch::ones({1, 3, 224, 224}).to(at::kCUDA));`
 `at::Tensor output = module->forward(inputs).toTensor();`
 `std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << '\n';`

get right reslut , all thing look like so perfect.
**but when I use my unet to Instead of resnet, the error occur.it will accident exit without any useful message.only show "Not loaded kerenlbase.pdb"**

I just use this code to convert my **UNET** model.
 `import torch `
 `import torch.nn.functional as F `
 `from unet import UNet `

 `model = UNet(n_channels=3, n_classes=1) `
 `model_path = "./MODEL.pth" `
 `model.load_state_dict(torch.load(model_path)) `
 `model.eval()`

 `example = torch.zeros(1, 3, 224, 224) `

 `traced_script_module = torch.jit.trace(model, example) `
 `traced_script_module.save("model.pt") `

**convert unet model is a little bit different of resnet.  one load from model zoo,other load by my local network and model.**
**cpp is exactly the same.**

Who can explain it? thank you.

**Comment by hset911** [comment]
> Could you please share the cpp code it generated and do some basic debugging like inserting some print clauses and find which one is causing the crash?

my cpp complete code :
`#include "torch/script.h"`
`#include "torch/torch.h"`

`#include <iostream>`
`#include <memory>`

`using namespace std;`

`int main()`
`{`
	`std::string model_file = "F:\\work\\Pytorch-UNet\\build\\Release\\model.pt";`
	`torch::DeviceType device_type = at::kCUDA;`
	`std::cout << "use cuda" << std::endl;`
    `std::shared_ptr<torch::jit::script::Module> module = torch::jit::load(model_file);`
    `std::cout << "load ok\n";`
    `module->to(device_type);`
    `assert(module != nullptr);`
    `std::cout << "ok\n";`
    `std::vector<torch::jit::IValue> inputs;`
    `inputs.push_back(torch::ones({1, 3, 224, 224}).to(at::kCUDA));`
    `at::Tensor output = module->forward(inputs).toTensor();`
    `std::cout << output.slice(/*dim=*/1, /*start=*/0, /*end=*/5) << '\n';`
`}`

It is certain that will not cout the **"load ok"**, So  it crash in **torch::jit::load()**.and without any other message.
it's belong to  libtorch'  interface function,......I don't have too much methods to debug it. feel helpless.

**Comment by peterjc123** [comment]
Would you please try to build with the nightly libtorch and see if it works? And there's one more question, are you building with the x64 Release configuration?

**Comment by DavidWongEA** [comment]
Is there a prebuild libtorch win debug version with pdb available?

**Comment by peterjc123** [comment]
@DavidWongEA Sorry, that is not available currently. I'll try to add it when I have time.

**Comment by hset911** [comment]
@peterjc123 I've tried to use nightly libtorch,it have same problem. and  I'm sure I used x64 Release.

**Comment by peterjc123** [comment]
@hset911 Sorry, I didn't know how the jit functions work. But I'll try to figure out the debug build that may provide some useful info.

**Comment by hset911** [comment]
@peterjc123 If you want to reproduce the bug that I described, I use the **pytorch_unet**:[https://github.com/milesial/Pytorch-UNet](url).
the project have network py file and pretrained model. 
only need to import it's network py file ,and download it model to try it. 
I hope you can find out the problem as soon as possible and fix it.
thank you.

**Comment by kdethoor** [comment]
Hi,
I encountered a similar problem today. I am using a TorchScript version of facebookresearch/maskrcnn-benchmark (serialized according to facebookresearch/maskrcnn-benchmark#138). In Python, the module can be loaded and used without any problem while in C++ it can't even be loaded: when `std::shared_ptr<torch::jit::script::Module> module = torch::jit::load(model_path);` is executed, the program silently stops/crashes.

Config:
* OS: Win10
* Arch: x64
* Device: CPU
* Pytorch version; 1.0 nightly (2019-01-14)

Let me know if I can provide anything else that could help.

**Comment by DavidWongEA** [comment]
@peterjc123 Thanks Peter! A prebuild pdb version would save many peoples time for quick debugging. 

For my case running on Window (which might be similar to @hset911), it cashes at top level (the sample example on your site) on this line:

   at::Tensor output = module->forward(inputs).toTensor();

Internally crashed at constants.cpp:

c10::optional<IValue> toIValue(const Value* v) {
  if (v->node()->kind() != prim::Constant) {
    return c10::nullopt;
  }
  // use implemenation of prim::Constant to compute the output IValue
  auto op = getOperation(v->node());
  Stack stack;
  op(stack);
  return stack.back();                                   <------------ here (due to the stack is empty)
}

Callstack:
>	torch.dll!torch::jit::toIValue(const torch::jit::Value * v) Line 164	C++
 	torch.dll!torch::jit::Node::get(c10::Symbol name) Line 640	C++
 	torch.dll!torch::jit::Node::is_constant(c10::Symbol name) Line 368	C++
 	torch.dll!torch::jit::Node::matches(const char * signature_literal, c10::ArrayRef<c10::Symbol> const_inputs) Line 653	C++
 	torch.dll!torch::jit::PeepholeOptimizeImpl(torch::jit::Block * block, bool addmm_fusion_enabled) Line 65	C++
 	torch.dll!torch::jit::PeepholeOptimize(torch::jit::Block * block, bool addmm_fusion_enabled) Line 175	C++
 	torch.dll!torch::jit::PeepholeOptimize(const std::shared_ptr<torch::jit::Graph> & graph, bool addmm_fusion_enabled) Line 182	C++
 	torch.dll!torch::jit::GraphExecutorImpl::runOptimization(std::shared_ptr<torch::jit::Graph> & graph, const torch::jit::ArgumentSpec & spec) Line 503	C++
 	torch.dll!torch::jit::GraphExecutorImpl::compileSpec(const torch::jit::ArgumentSpec & spec) Line 473	C++
 	torch.dll!torch::jit::GraphExecutorImpl::getOrCompile(const std::vector<c10::IValue,std::allocator<c10::IValue> > & stack) Line 440	C++
 	torch.dll!torch::jit::GraphExecutorImpl::run(std::vector<c10::IValue,std::allocator<c10::IValue> > & stack) Line 374	C++
 	torch.dll!torch::jit::GraphExecutor::run(std::vector<c10::IValue,std::allocator<c10::IValue> > & inputs) Line 610	C++
 	example-app.exe!torch::jit::script::Method::run(std::vector<c10::IValue,std::allocator<c10::IValue> > & stack) Line 72	C++
 	example-app.exe!torch::jit::script::Method::operator()(std::vector<c10::IValue,std::allocator<c10::IValue> > stack) Line 77	C++
 	example-app.exe!torch::jit::script::Module::forward(std::vector<c10::IValue,std::allocator<c10::IValue> > inputs) Line 384	C++
 	example-app.exe!main(int argc, const char * * argv) Line 45	C++

It's real cause is due to the line above it op(stack). 
The lambda operator returned which supposed to do a push onto the stack but failed (in constant.cpp):

           return [i](Stack& stack) {
              push(stack, i);
              return 0;
            };

The push function doesn't work for the above case where only pushing one element on the stack. Pushing more elements like push(stack, 1, 2) worked fine. But not push(stack, 1). Probably due to a combination of this particular implementation and VS compiler (VS 15.6.6).

In jit\stack.h:

template <typename... Types>
static inline void push(Stack& stack, Types&&... args) {
  constexpr size_t N = sizeof...(args);
  int result[N] = {(stack.emplace_back(std::forward<Types>(args)), 0)...};
  (void)result;
}

My colleague Evan who's an expert in c11 suggested trying to change the implementation to the following:

template <typename... Types>
static inline void push(Stack& stack, Types&&... args) {
   std::initializer_list<int>{(stack.emplace_back(std::forward<Types>(args)), 0)...};
}

Which looks cleaner and worked great on Window! 
So posted here, in case, it's a potential fix for others.

**Comment by peterjc123** [comment]
@DavidWongEA Thanks for letting me know. Would you please make a PR on this?

**Comment by marcelpohl** [comment]
Hi,
is there any suggestion which version of Visual Studio to use if it's a compiler problem?
I got my projects to build, but every example I've tried so far (https://pytorch.org/cppdocs/frontend.html and https://pytorch.org/tutorials/advanced/cpp_export.html) both just crash in the very beginning. 

- When creating the net in C++ the program crashes when instantiating  (auto net = std::make_shared<Net>();  ) throwing a "vector subscript out of range" error
- When trying to load a model  (std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("test.pt");  )  it throws the exeption: c10::Error at memory location 0x000.....

Currently I'm working on Win10 with Visual Studio 2017 (15.9.4). I've tried the stable and nightly build, using x64 (it wouldn't even build otherwise). Building for debug or release, it always results in the same crashes when running the examples. When creating just simple Tensors they do work, but it's no good when the rest doesn't. Is there any configuration that is confirmed to work well? I mean, someone has to have used it some time before it was released ...

**Comment by peterjc123** [comment]
@marcelpohl Oh, I don't know why you post it here. Apparently, they are different issues. This one is for jit and yours is about libtorch and cpp frontend. And actually, the answers could be found at previous issues if you could do some search. Anyway, I'll answer it again.
1. It is expected. Because in VS, debug and release configuration could not be mixed, especially when we use something like `std::string` or `std::vector`. Please use Release because we only provide the Release version of LibTorch.
2. Are you building with VS GUI? If yes, please check if the `data` directory (created by CMake) or the `model.pt` locates at the same directory with the executable? Suggestion: If you use VS GUI to build CMake projects, then please use the CMake integration. In that way, this issue won't happen again.
Additional words: Please prefer the nightly LibTorch because it is not stable enough even in the Stable build at present.

**Comment by marcelpohl** [comment]
@peterjc123 Finally got it to run, thanks. I did read about the release only build, and tried it when trying to load the model I saved in python, and apparently forgot about it when trying the MNIST example. 

Also, I replied to this topic since I essentially need to load models too, and I didn't know if I was having the same problem or not. I don't know if there is an alternative to save and load models using purely C++, I'll check out the docs. Anyways, I'll check the nightly build regularly now to see when the potential fix is included and will try again, thanks for your help.

**Comment by suo** [comment]
Alright, going to close this out as it seems the various issues raised have been resolved. For future users, we're adding a tip to the tutorial noting the debug/release incompatibility. The workaround is to either build your project in release mode or build libtorch from source.
